zilencer/views.py

```diff
@@ -150,15 +150,28 @@ def remote_server_notify_push(request: HttpRequest, entity: Union[UserProfile, R
 
     return json_success()
 
-def validate_count_stats(server: RemoteZulipServer, model: Any,
-                         counts: List[Dict[str, Any]]) -> None:
+def validate_incoming_table_data(server: RemoteZulipServer, model: Any,
+                                 rows: List[Dict[str, Any]], is_count_stat: bool=False) -> None:
     last_id = get_last_id_from_server(server, model)
-    for item in counts:
-        if item['property'] not in COUNT_STATS:
-            raise JsonableError(_("Invalid property %s") % (item['property'],))
-        if item['id'] <= last_id:
+    for row in rows:
+        if is_count_stat and row['property'] not in COUNT_STATS:
+            raise JsonableError(_("Invalid property %s") % (row['property'],))
+        if row['id'] <= last_id:
             raise JsonableError(_("Data is out of order."))
-        last_id = item['id']
+        last_id = row['id']
+
+def batch_create_table_data(server: RemoteZulipServer, model: Any,
+                            row_objects: Union[List[RemoteRealmCount],
+                                               List[RemoteInstallationCount]]) -> None:
+    BATCH_SIZE = 1000
+    while len(row_objects) > 0:
+        try:
+            model.objects.bulk_create(row_objects[:BATCH_SIZE])
+        except IntegrityError:
+            logging.warning("Invalid data saving %s for server %s/%s" % (
+                model._meta.db_table, server.hostname, server.uuid))
+            return json_error(_("Invalid data."))
```

Comment:
The refactor moving this out of the parent broke things; should be `raise JsonableError` now.  Fortunately, we had a unit test for this case, which failed.  I just fixed that and will merge this commit with that fixed.